nupic/research/frameworks/vernon/mixins/track_representation_sparsity.py

```diff
@@ -0,0 +1,252 @@
+from nupic.research.frameworks.pytorch.model_utils import (
+    filter_modules,
+    get_module_attr,
+)
+
+
+class TrackRepresentationSparsity:
+    """
+    This mixin tracks and reports the average sparsities observed in the model's
+    representations. Either input or output sparsities can be tracked for a specified,
+    possibly overlapping, subset of modules. Tracked statistics are returned from
+    `run_epoch` and are reset before each subsequent epoch. The default is to track none
+    of the modules; only those specified can be tracked.
+
+    .. _filter_modules: nupic.research.frameworks.pytorch.model_utils.filter_modules
+    """
+
+    def setup_experiment(self, config):
+        """
+        Register forward hooks to track the input and output sparsities.
+
+        The subsets of modules to track are defined via `include_*` params. See
+        `filter_modules`_ for further details.
+
+        :param config:
+            - track_input_sparsity_args:
+                - include_modules: a list of module types to track
+                - include_names: a list of module names to track e.g. "features.stem"
+                - include_patterns: a list of regex patterns to compare to the names;
+                                    for instance, all feature parameters in ResNet can
+                                    be included through "features.*"
+            - track_output_sparsity_args: same as track_input_sparsity_args
+
+        .. _filter_modules: nupic.research.frameworks.pytorch.model_utils.filter_modules
+        """
+        super().setup_experiment(config)
+
+        # Register hooks to track input sparsities.
+        input_tracking_args = config.get("track_input_sparsity_args", {})
+        named_modules = filter_modules(self.model, **input_tracking_args)
+        self.input_hook_manager = ModelHookManager(named_modules, TrackSparsityHook)
+
+        # Log the names of the modules with tracked inputs.
+        tracked_names = pformat(list(named_modules.keys()))
+        self.logger.info(f"Tracking input sparsity of modules: {tracked_names}")
+
+        # Register hooks to track output sparsities.
+        output_tracking_args = config.get("track_output_sparsity_args", {})
+        named_modules = filter_modules(self.model, **output_tracking_args)
+        self.output_hook_manager = ModelHookManager(named_modules, TrackSparsityHook)
+
+        # Log the names of the modules with tracked outputs.
+        tracked_names = pformat(list(named_modules.keys()))
+        self.logger.info(f"Tracking output sparsity of modules: {tracked_names}")
+
+        # Throw a warning when no modules are being tracked.
+        if not input_tracking_args and not output_tracking_args:
+            self.logger.warning("No modules specified to track input/output sparsity.")
+
+    def run_epoch(self):
+        """Run one epoch and log the observed sparsity statistics."""
+
+        with self.input_hook_manager, self.output_hook_manager:
```

Comment:
can you explain what's the effect of this line? `self.input_hook_manager` is of type `ModelHookManager`, but I don't follow what this `with` clause is doing.

Reply:
This `with` is invoking them as context managers. You're familiar with `open` as a context manager where a `with open(...):` block opens and closes a file. Another typical use case for context managers is to start and stop some process (e.g. `with torch.no_grad()`).

If you recall, the `ModelHookManager` has methods `__enter__` and `__exit__`. These define the behavior of the class upon entering and exiting a `with` block. Respectively, the manager starts and stops tracking at these points.

Reply:
Thus, the following are equivalent.
```
hook_manager.start_tracking()
... # do some training
hook_manager_stop_tracking()
stats = hook_manager.get_statistics()
```
and
```
with hook_manager:
... # do some training
stats = hook_manager.get_statistics()
```